src/connected_component.rs

```diff
@@ -0,0 +1,215 @@
+use std::collections::{ HashSet };
+use weighted_graph::{ Graph, NodeId };
+use dijkstra::shortest_path;
+
+pub fn reduce_to_largest_connected_component(graph: Graph) -> Graph {
+    let untested_nodes = node_ids(&graph);
+    reducer(graph, untested_nodes, vec![])
+}
+
+fn reducer(graph: Graph, untested_nodes: HashSet<NodeId>, mut results: Vec<HashSet<NodeId>>) -> Graph {
+    if untested_nodes.is_empty() {
+        collapsed_graph(&graph, &results)
+    } else {
+        let root = untested_nodes.iter().nth(0).unwrap();
+        let connected_nodes = explore_from(&root, &graph);
+        results.push(connected_nodes);
+        reducer(graph,
+                untested_nodes.difference(results.last().unwrap()).cloned().collect(),
+                results
+                )
+    }
+}
+
+fn explore_from(root: &NodeId, graph: &Graph) -> HashSet<NodeId> {
+    let (_, results) = shortest_path(graph, root, None);
+    results.values()
+           .map(|result| result.id.clone())
+           .collect::<HashSet<NodeId>>()
+}
+
+fn collapsed_graph(graph: &Graph, results: &Vec<HashSet<NodeId>>) -> Graph {
+    let mut new_graph = Graph::new();
+    match results.iter().max_by_key(|nodes| nodes.len()) {
+        Some(nodes) => {
+            for node_id in nodes {
+                add_node(graph, &mut new_graph, &node_id);
+            }
+            for node_id in nodes {
+                add_edges(graph, &mut new_graph, &node_id);
+            }
```

Comment:
The nodes are all added first, and then the edges are added. The way I originally implemented the graph, and edge was not added if the from and to nodes were not both present, so I need to add all the nodes before I add any of the edges. `add_edges` doesn't add any nodes; it looks up the edges for each node in the original graph, and copies all the edges from that node to the new graph (for every edge where the `to_node` also exists).